File: api/internal/domain/message.go
package domain

import (
	"errors"
	"time"
)

var (
	ErrInvalidMessageContent = errors.New("message content must be 1-2048 characters")
	ErrInvalidMessageType    = errors.New("invalid message type")
	ErrInvalidSOSType        = errors.New("invalid SOS type")
	ErrSOSTypeRequired       = errors.New("SOS type is required for SOS messages")
)

// MessageType represents the type of a message
type MessageType string

const (
	MessageTypeText         MessageType = "text"
	MessageTypeSOS          MessageType = "sos"
	MessageTypeStatusUpdate MessageType = "status_update"
)

// SOSType represents the type of SOS emergency
type SOSType string

const (
	SOSTypeMedical       SOSType = "medical"
	SOSTypeFlood         SOSType = "flood"
	SOSTypeFire          SOSType = "fire"
	SOSTypeMissingPerson SOSType = "missing_person"
)

// Message represents a communication within a group
type Message struct {
	ID             string      `json:"id"`
	GroupID        string      `json:"group_id"`
	DeviceID       string      `json:"device_id"`
	Content        string      `json:"content"`
	MessageType    MessageType `json:"message_type"`
	SOSType        *SOSType    `json:"sos_type,omitempty"`
	Tags           []string    `json:"tags,omitempty"`
	Pinned         bool        `json:"pinned"`
	CreatedAt      time.Time   `json:"created_at"`
	DeviceSequence *int        `json:"device_sequence,omitempty"`
	SyncedAt       *time.Time  `json:"synced_at,omitempty"`
}

// Validate validates message fields
func (m *Message) Validate() error {
	// Basic content length guard (1–2048 characters, ~2KB max)
	if len(m.Content) < 1 || len(m.Content) > 2048 {
		return ErrInvalidMessageContent
	}
	// Basic group and device validation
	if m.GroupID == "" {
		return errors.New("group_id is required")
	}
	if m.DeviceID == "" {
		return errors.New("device_id is required")
	}

	// Validate tags: limit count and length
	if len(m.Tags) > 16 {
		return errors.New("too many tags (max 16)")
	}
	for _, tag := range m.Tags {
		if len(tag) == 0 {
			return errors.New("tag cannot be empty")
		}
		if len(tag) > 64 {
			return errors.New("tag too long (max 64 characters)")
		}
	}

	if !m.MessageType.IsValid() {
		return ErrInvalidMessageType
	}
	if m.MessageType == MessageTypeSOS {
		if m.SOSType == nil {
			return ErrSOSTypeRequired
		}
		if !m.SOSType.IsValid() {
			return ErrInvalidSOSType
		}
	} else if m.SOSType != nil {
		return errors.New("SOS type should only be set for SOS messages")
	}
	return nil
}

// IsValid checks if MessageType is valid
func (mt MessageType) IsValid() bool {
	switch mt {
	case MessageTypeText, MessageTypeSOS, MessageTypeStatusUpdate:
		return true
	default:
		return false
	}
}

// IsValid checks if SOSType is valid
func (st SOSType) IsValid() bool {
	switch st {
	case SOSTypeMedical, SOSTypeFlood, SOSTypeFire, SOSTypeMissingPerson:
		return true
	default:
		return false
	}
}

File: web/src/shared/domain/message.ts
/**
 * Message domain model
 * Represents a communication within a group
 */

export type MessageType = 'text' | 'sos' | 'status_update';

export type SOSType = 'medical' | 'flood' | 'fire' | 'missing_person';

export type SyncStatus = 'pending' | 'syncing' | 'synced' | 'failed';

export interface Message {
  id: string; // NanoID (21 chars)
  group_id: string; // NanoID (21 chars)
  device_id: string; // NanoID (21 chars)
  content: string; // 1-500 characters
  message_type: MessageType;
  sos_type?: SOSType; // Required if message_type is 'sos'
  tags?: string[]; // Optional tags
  pinned: boolean;
  created_at: string; // ISO timestamp
  device_sequence?: number; // Device-local sequence number
  synced_at?: string; // ISO timestamp
  sync_status?: SyncStatus; // Client-side sync status
}

export interface MessageCreateRequest {
  group_id: string; // NanoID (21 chars)
  content: string; // 1-500 characters
  message_type: MessageType;
  sos_type?: SOSType; // Required if message_type is 'sos'
  tags?: string[];
  device_sequence?: number;
}

/**
 * Validates message type
 * @param type - Message type to validate
 * @returns True if valid
 */
export function isValidMessageType(type: string): type is MessageType {
  return ['text', 'sos', 'status_update'].includes(type);
}

/**
 * Validates SOS type
 * @param type - SOS type to validate
 * @returns True if valid
 */
export function isValidSOSType(type: string): type is SOSType {
  return ['medical', 'flood', 'fire', 'missing_person'].includes(type);
}

/**
 * Validates message content
 * @param content - Message content to validate
 * @returns Error message if invalid, null if valid
 */
export function validateMessageContent(content: string): string | null {
  if (content.length < 1 || content.length > 500) {
    return 'Message content must be 1-500 characters';
  }
  // Check for control characters (except newline)
  // eslint-disable-next-line no-control-regex
  const controlCharPattern = /[\u0000-\u0008\u000B-\u000C\u000E-\u001F\u007F]/;
  if (controlCharPattern.test(content)) {
    return 'Message content contains invalid characters';
  }
  return null;
}

/**
 * Validates message data
 * @param message - Message to validate
 * @returns Error message if invalid, null if valid
 */
export function validateMessage(message: Message): string | null {
  const contentErr = validateMessageContent(message.content);
  if (contentErr) return contentErr;
  if (!isValidMessageType(message.message_type)) return 'Invalid message type';
  
  if (message.message_type === 'sos') {
    if (!message.sos_type) {
      return 'SOS type is required for SOS messages';
    }
    if (!isValidSOSType(message.sos_type)) {
      return 'Invalid SOS type';
    }
  } else if (message.sos_type) {
    return 'SOS type should only be set for SOS messages';
  }
  
  return null;
}

/**
 * Validates message creation request
 * @param request - Message creation request to validate
 * @returns Error message if invalid, null if valid
 */
export function validateMessageCreateRequest(request: MessageCreateRequest): string | null {
  const contentErr = validateMessageContent(request.content);
  if (contentErr) return contentErr;
  if (!isValidMessageType(request.message_type)) return 'Invalid message type';
  
  if (request.message_type === 'sos') {
    if (!request.sos_type) {
      return 'SOS type is required for SOS messages';
    }
    if (!isValidSOSType(request.sos_type)) {
      return 'Invalid SOS type';
    }
  } else if (request.sos_type) {
    return 'SOS type should only be set for SOS messages';
  }
  
  return null;
}

